crates/bevy_ecs/src/query/access.rs

```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrAccess<T, I> {
+    arr: SmallVec<[FixedBitSet; 8]>,
+    // To prevent mixing `With` and `Without` filters.
+    _filter_type: PhantomData<T>,
+    _index_type: PhantomData<I>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithMarker;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithoutMarker;
+
+impl<T, I: SparseSetIndex + fmt::Debug> fmt::Debug for ExpandedOrAccess<T, I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list()
+            .entries(self.arr.iter().map(FormattedBitSet::<I>::new))
+            .finish()
+    }
+}
+
+impl<T, I: SparseSetIndex> Default for ExpandedOrAccess<T, I> {
+    fn default() -> Self {
+        Self {
+            arr: smallvec::smallvec![FixedBitSet::default()],
+            _filter_type: PhantomData,
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T, I: SparseSetIndex> ExpandedOrAccess<T, I> {
```

Comment:
Each of these methods needs to be carefully commented. This is tricky soundness critical code, and I'd like to see the logic for each more explicitly.

Reply:
(the implementation evolved a bit, and so did the comments, resolving it in favour of comments to the up-to-date diff)